includes/class-events-store.php

```diff
@@ -882,6 +879,315 @@ public function count_events_by_status( $status ) {
+		$placeholders = [];
+
+		// Timestamp can be:
+		if ( ! is_null( $parsed_args['timestamp'] ) ) {
+			// 1) A direct integer.
+			if ( is_int( $parsed_args['timestamp'] ) ) {
+				$sql .= ' AND timestamp = %d';
+				$placeholders[] = $parsed_args['timestamp'];
+			}
+
+			// 2) Or a request for everything that is "due now".
+			if ( 'due_now' === $parsed_args['timestamp'] ) {
+				$sql .= ' AND timestamp <= %d';
+				$placeholders[] = time();
+			}
+
+			// 3) Or a range between two timestamps.
+			if ( is_array( $parsed_args['timestamp'] ) ) {
+				$sql .= ' AND timestamp >= %d AND timestamp <= %d';
+				$placeholders[] = $parsed_args['timestamp']['from'];
+				$placeholders[] = $parsed_args['timestamp']['to'];
+			}
+		}
+
+		if ( ! is_null( $parsed_args['action'] ) ) {
+			$sql .= ' AND action = %s';
+			$placeholders[] = $parsed_args['action'];
+		}
+
+		if ( ! is_null( $parsed_args['action_hashed'] ) ) {
+			// TODO: Deprecate this query arg later once all is converted to the new API.
+			$sql .= ' AND action_hashed = %s';
+			$placeholders[] = $parsed_args['action_hashed'];
+		}
+
+		if ( ! is_null( $parsed_args['args'] ) ) {
+			// Rather than query args directly, convert to the hash so we can utilize index.
+			$instance = Event::create_instance_hash( $parsed_args['args'] );
+			$sql .= ' AND instance = %s';
+			$placeholders[] = $instance;
+		} elseif ( ! is_null( $parsed_args['instance'] ) ) {
+			// TODO: Deprecate this query arg later once all is converted to the new API.
+			$sql .= ' AND instance = %s';
+			$placeholders[] = $parsed_args['instance'];
+		}
+
+		if ( ! is_null( $parsed_args['schedule'] ) ) {
+			$sql .= ' AND schedule = %s';
+			$placeholders[] = $parsed_args['schedule'];
+		}
+
+		$requested_any_status = is_string( $parsed_args['status'] ) ? 'any' === strtolower( $parsed_args['status'] ) : false;
+		if ( ! $requested_any_status ) {
+			if ( is_array( $parsed_args['status'] ) ) {
+				$statuses = array_map( 'strtolower', $parsed_args['status'] );
+				$sql .= ' AND status IN (' . implode( ',', array_fill( 0, count( $statuses ), '%s' ) ) . ')';
+				$placeholders = array_merge( $placeholders, $statuses );
+			} elseif ( is_string( $parsed_args['status'] ) ) {
+				$sql .= ' AND status = %s';
+				$placeholders[] = strtolower( $parsed_args['status'] );
+			}
+		}
+
+		// TODO: adjust for situations where we don't need to sort.
+		$sql .= ' ORDER BY timestamp';
+		$sql .= strtoupper( $parsed_args['order'] ) === 'ASC' ? ' ASC' : ' DESC';
+
+		// Skip paging/limits if "-1" was passed to get all events.
+		if ( $parsed_args['limit'] >= 1 ) {
+			$sql .= ' LIMIT %d';
+			$placeholders[] = $parsed_args['limit'];
+
+			if ( ! is_null( $parsed_args['page'] ) ) {
+				$offset = $parsed_args['limit'] * ( $parsed_args['page'] - 1 );
+				if ( $offset > 0 ) {
+					$sql .= ' OFFSET %d';
+					$placeholders[] = $offset;
+				}
+			}
+		}
+
+		$last_changed = wp_cache_get_last_changed( 'cron-control-events' );
```

Comment:
You were right! This is kinda smart! :)

Reply:
Somehow I never picked up on the core patterns that use this. Makes cache groups a lot more useful. https://github.com/WordPress/WordPress/blob/121c824ccde0a020f2e92b6d287d3a6fc35a913f/wp-includes/post.php#L5630-L5634